turbopack/crates/turbo-persistence/src/static_sorted_file_builder.rs

```diff
@@ -94,387 +94,472 @@ pub struct StaticSortedFileBuilderMeta<'a> {
-                    let used_len = min(value_remaining, COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY);
-                    if len <= used_len {
-                        value_sample_sizes.push(len);
-                        value_samples.extend_from_slice(value);
-                    } else {
-                        value_sample_sizes.push(used_len);
-                        let p = value_samples.len() % (len - used_len);
-                        value_samples.extend_from_slice(&value[p..p + used_len]);
-                    };
-                }
-            }
+        size = size.max(value_compression_samples_size);
+    }
+    buffer.reserve(size);
+}
+
+/// Computes compression dictionaries from keys of all entries
+#[tracing::instrument(level = "trace", skip(entries))]
+fn compute_key_compression_dictionary<E: Entry>(
+    entries: &[E],
+    total_key_size: usize,
+    buffer: &mut Vec<u8>,
+) -> Result<Vec<u8>> {
+    if total_key_size < MIN_KEY_COMPRESSION_SAMPLES_SIZE {
+        return Ok(Vec::new());
+    }
+    let key_compression_samples_size = min(KEY_COMPRESSION_SAMPLES_SIZE, total_key_size / 16);
+    buffer.reserve(key_compression_samples_size);
+    let mut sample_sizes = Vec::new();
+
+    // Limit the number of iterations to avoid infinite loops
+    let max_iterations = total_key_size / COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY * 2;
+    for i in 0..max_iterations {
+        let entry = &entries[i % entries.len()];
+        let key_remaining = key_compression_samples_size - buffer.len();
+        if key_remaining < MIN_COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY {
+            break;
         }
-        assert!(value_samples.len() == value_sample_sizes.iter().sum::<usize>());
-        if value_samples.len() > MIN_VALUE_COMPRESSION_SAMPLES_SIZE && value_sample_sizes.len() > 5
-        {
-            self.value_compression_dictionary = zstd::dict::from_continuous(
-                &value_samples,
-                &value_sample_sizes,
-                VALUE_COMPRESSION_DICTIONARY_SIZE,
-            )
-            .context("Value dictionary creation failed")?;
-        } else {
-            self.value_compression_dictionary = Vec::new();
+        let len = entry.key_len();
+        if len >= MIN_COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY {
+            let used_len = min(key_remaining, COMPRESSION_DICTIONARY_SAMPLE_PER_ENTRY);
+            if len <= used_len {
+                sample_sizes.push(len);
+                entry.write_key_to(buffer);
+            } else {
+                let mut temp = Vec::with_capacity(len);
+                entry.write_key_to(&mut temp);
+                debug_assert!(temp.len() == len);
+
+                let p = buffer.len() % (len - used_len);
+                sample_sizes.push(used_len);
+                buffer.extend_from_slice(&temp[p..p + used_len]);
+            }
         }
+    }
+    assert!(buffer.len() == sample_sizes.iter().sum::<usize>());
```

Comment:
Should this be a `debug_assert`? The `sum` seems potentially expensive?

```suggestion
    debug_assert!(buffer.len() == sample_sizes.iter().sum::<usize>());
```